### PR TITLE
TK-06685 TK-06875 TK-06673 remove all explicit lifetimes and references in hdk signatures

### PR DESCRIPTION
**Pull request by thedavidmeister**
removes all lifetimes and references in the HDK and in the process fixes the issue where hashing an `Entry` and the internal data gives two different results

### CODE REVIEW

**Comment by thedavidmeister** [comment]
i can see the problem:

```
test_wasm_hash_entry:hash_entry/src/lib.rs:25 EntryHash(uhCEkAw11yvUilWRbxhNrer8-HlwlpZHGzWNkFK2Dr9wSvwe33yqd)
[crates/holochain/src/core/ribosome/host_fn/hash_entry.rs:85] &output = Element {
    signed_header: SignedHeaderHashed {
        header: HoloHashed(Create(Create { author: AgentPubKey(uhCAkmrkoAHPVf_eufG7eC5fm6QKrW5pPMoktvG5LOC0SnJ4vV1Uv), timestamp: Timestamp(1609786395, 887817680), header_seq: 3, prev_header: HeaderHash(uhCkktcjOp2xYxMYxzXUgPex1Px4xYGtUQzlJIwQHyiiGux42GYnr), entry_type: App(AppEntryType { id: EntryDefIndex(0), zome_id: ZomeId(0), visibility: Public }), entry_hash: EntryHash(uhCEkiBQDwRYA7pqETw1kvZ-XyU28Embey18fYost4Znvlg7CcPle) })),
        signature: Signature(0xae0fe54eab8850175d8d285b3ad38cf07be4ff6ff44a84cf91e31a43f7ad7d75013853208603808f5e47b062df37fa724d2cf11e5919f1af0ba238ab95fbb10b),
    },
    entry: Present(
        App(
            AppEntryBytes(
                [32,{"Celcius":null}],
            ),
        ),
    ),
}
[crates/holochain/src/core/ribosome/host_fn/hash_entry.rs:88] &hash_output = EntryHash(uhCEkiBQDwRYA7pqETw1kvZ-XyU28Embey18fYost4Znvlg7CcPle)
```

the first hash is the result of entry hashing the `Entry::App`

File: crates/hdk3/src/guest_callback/entry_defs.rs [hunk 1]
```diff
@@ -84,6 +84,20 @@ macro_rules! entry_def {
...
+            type Error = $crate::prelude::HdkError;
+            fn try_from(t: &$t) -> Result<Self, Self::Error> {
+                Ok(Self::App(SerializedBytes::try_from(t)?.try_into()?))
+            }
+        }
+
+        impl TryFrom<$t> for $crate::prelude::Entry {
+            type Error = $crate::prelude::HdkError;
+            fn try_from(t: $t) -> Result<Self, Self::Error> {
+                Self::try_from(&t)
```
Comment (freesig) [inline crates/hdk3/src/guest_callback/entry_defs.rs:97]: What happens if they already have an`Entry` does this result in recursively serialising it?

**Review by freesig — CHANGES_REQUESTED** [review]
Looks good, I'm not sure what happens if you pass in an `Entry` I think it will result in reserializing?
Maybe you want to implement try from `&Entry` for `Entry` and just clone it?

File: crates/hdk3/src/guest_callback/entry_defs.rs (hunk 1 above)
Comment (thedavidmeister) [reply crates/hdk3/src/guest_callback/entry_defs.rs:97]: @freesig no because you don't use `entry_defs!` macro on `Entry`

**Review by freesig — APPROVED** [review]
Actually I get it now 👍

**Comment by pospi** [comment]
When I want to use a higher-order method which both hashes and creates an entry, I notice that I need two trait bounds.

One targets `holochain_zome_types::entry::Entry`, the other targets `hdk3::entry::HdkEntry`. Presumably these are different types because they exist on different sides of the WASM boundary, but I still find this a curious quirk.

Can you explain what the difference between the two is, and does this also mean that I need to parameterise raw entry data somehow to indicate that it's an app entry before passing to `hdk3::prelude::hash_entry`? Or do the conversion traits abstract away that detail?

**Comment by pospi** [comment]
This also, as far as I can tell, means that you now have to `.clone()` data when calling multiple HDK core methods in the same function:

```rust
pub fn create_entry<I, E>(
    entry_struct: I,
) -> GraphAPIResult<(HeaderHash, EntryHash)>
    where HdkError: From<E>,
        HdkEntry: TryFrom<I, Error = E>,
        Entry: TryFrom<I, Error = E>,
{
    let entry_hash = hash_entry(entry_struct)?;
    let header_hash = hdk_create_entry(entry_struct)?;  // use of moved value: `entry_struct`

    Ok((header_hash, entry_hash))
}
```

**Pull request merged**